text/0000-posix-error-numbers.md

```diff
@@ -0,0 +1,248 @@
+
+The POSIX standard defines a set of named [error numbers], such as `EIO` or
+`EPIPE`, for use in communicating error conditions from low-level OS routines.
+Many of these are currently exposed in [`libc`] as `c_int` constants, because
+they figure prominently in writing bindings to libc on UNIX platforms.
+
+However, there are two problems with this solution:
+
+* POSIX does not specify the type of these error numbers -- they are defined as
+  C macros, which are untyped integer literals. Many traditional UNIX APIs
+  assume that they can use any integer type for returning error numbers, which
+  are unergonomic to compare with a `c_int` const.
+
+* Binaries that interact with the kernel directly (e.g. via `asm!()` syscalls)
+  need to interpret error numbers, and should not require a dependency on libc
+  to do so.
+
+I propose adding an opaque `ErrorNumber` struct to `std::os::unix`, which would
+provide limited conversion and comparison capabilities. By restricting the API
+surface it should be possible to provide similar ergonomics to untyped integer
+literals without abandoning type safety.
+
+[error numbers]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_03
+[`libc`]: https://crates.io/crates/libc
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The POSIX standard defines a set of named [error numbers], such as `EPERM` or
+`ENOENT`, for use in communicating error conditions from low-level OS routines.
+POSIX does not specify the type or value of error numbers, so they vary between
+operating systems and architectures. Many traditional UNIX APIs assume that
+error numbers may be converted to or compared with any numeric value.
+
+The `std::os::unix::ErrorNumber` struct represents an abstract non-zero error
+number, which can be compared with any of the built-in integral types.
+
+```rust
+use std::os::raw::c_long;
+use std::os::unix::ErrorNumber;
+
+fn some_libc_fn() -> Option<ErrorNumber> {
+  while {
+    let err: c_long = libc::some_fn();
+    if err == 0 {
+      return None;
+    }
+    if err != ErrorNumber::EAGAIN {
+      return Some(err);
+    }
+  }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are two broad categories of users for this type:
+
+* Binaries linking against `std` that call `libc` functions.
+* Binaries built with `#![no_std]` that invoke kernel syscalls.
+
+Since `ErrorNumber` is platform-specific functionality it doesn't belong in
+`core`, so it must be placed in a separate crate that can be depended on by
+`#![no_std]` binaries. `std` would then depend on this crate and re-export
+its symbols.
+
+The internal representation of `ErrorNumber` should be either `NonZeroI32` or
+`NonZeroI64` depending on the target architecture. This is for performance, to
+avoid sign-extending instructions when returning error numbers (e.g. in a tight
+loop, or up a long call stack).
```

Comment:
I don't understand how this would break the code in the link.

For functions that use the high bit to signal an error condition, the caller will need to know through some other means what the highest error number is -- for example in Linux it's 4096 -- so they can inspect the return code with bitmasks.